### DISCUSSION TITLE
How to call an API from a different API in the same app in FastAPI?

### DISCUSSION THREAD

**Discussion by kashyapm94** [fastapi/fastapi#6279] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
See description
```

### Description

(I did find the following question on github, but it didn't help me: https://github.com/tiangolo/fastapi/issues/199#issuecomment-492541834

I am making an app using Fastapi with the following folder structure

[![enter image description here][1]][1]

`main.py` is the entry point to the app

    from fastapi import FastAPI
    from fastapi.middleware.cors import CORSMiddleware
    
    from app.api.v1 import lines, upload
    from app.core.config import settings
    
    app = FastAPI(
        title=settings.PROJECT_NAME,
        version=0.1,
        openapi_url=f'{settings.API_V1_STR}/openapi.json',
        root_path=settings.ROOT_PATH
    )
    
    app.add_middleware(
        CORSMiddleware,
        allow_origins=settings.BACKEND_CORS_ORIGINS,
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )
    
    app.include_router(upload.router, prefix=settings.API_V1_STR)
    app.include_router(lines.router, prefix=settings.API_V1_STR)

In the `lines.py`, I have 2 GET endpoints:

 - `/one-random-line` --> returns a random line from a `.txt` file
 - `/one-random-line-backwards` --> should return the output of the `/one-random-line`

Since the output of the second GET endpoint should be the reversed string of the output of the first GET endpoint, I tried doing the following steps mentioned [here][2]

The codes:

    import random
    
    from fastapi import APIRouter, Request
    from starlette.responses import RedirectResponse
    
    router = APIRouter(
        prefix="/get-info",
        tags=["Get Information"],
        responses={
            200: {'description': 'Success'},
            400: {'description': 'Bad Request'},
            403: {'description': 'Forbidden'},
            500: {'description': 'Internal Server Error'}
        }
    )
    
    
    @router.get('/one-random-line')
    def get_one_random_line(request: Request):
        lines = open('netflix_list.txt').read().splitlines()
        if request.headers.get('accept') in ['application/json', 'application/xml']:
            random_line = random.choice(lines)
        else:
            random_line = 'This is an example'
        return {'line': random_line}
    
    
    @router.get('/one-random-line-backwards')
    def get_one_random_line_backwards():
        url = router.url_path_for('get_one_random_line')
        response = RedirectResponse(url=url)
        return {'message': response[::-1]}

When I do this, I get the following error:

    TypeError: 'RedirectResponse' object is not subscriptable

When I change the `return` of the second GET endpoint to `return {'message': response}`, I get the following output

[![enter image description here][3]][3]

What is the mistake I am doing?

**Example:**

If the output of `/one-random-line` endpoint is 'Maverick', then the output of `/one-random-line-backwards` should be 'kcirevam'

  [1]: https://i.stack.imgur.com/YPor6.png
  [2]: https://github.com/tiangolo/fastapi/issues/199#issuecomment-492541834
  [3]: https://i.stack.imgur.com/AYVPc.png
  [4]: https://stackoverflow.com/questions/70079723/is-it-possible-to-have-an-api-call-another-api-having-them-both-in-same-applica

### Operating System

Linux

### FastAPI Version

0.79.0

### Python Version

3.8

**Comment by iudeen** [comment]
Since it's a same Python code base, you can simply call it like a function. It's better you have a common handler of some sorts. 

Also, the error you get is because you're not returning the RedirectResponse object directly

**Comment by kashyapm94** [comment]
@iudeen What do you mean when you say 'have a common handler'? Could you maybe give an example?

**Comment by JarroVGIT** [comment, marked as answer]
You don't have to use any HTTP calls or anything, just call the endpoint directly like a function. See the below example, it will run as-is.

```python
from fastapi import FastAPI, Request

app = FastAPI()


@app.get("/one-random-line")
async def get_one_random_line(request: Request):
    # implement your own logic here, this will only return a static line
    return {"line": "This is an example"}


@app.get("/one-random-line-backwards")
async def get_one_random_line_backwards(request: Request):
    # You don't have to do fancy http stuff, just call your endpoint:
    one_line = await get_one_random_line(request)
    return {"line": one_line["line"][::-1]}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)

```

using curl:
```bash
% curl localhost:8000/one-random-line          
{"line":"This is an example"}%     
% curl localhost:8000/one-random-line-backwards
{"line":"elpmaxe na si sihT"}%         
```

**Reply by princyangelina** [reply]
Hi, how can I change the get_one_random_line_backwards() if I just want to call the get_one_random_line() in it and not wait for the response from it and return something else which doesn't depend on the other method?
